File: LeetCode/658.Find_K_Closest_Elements/658.Find_K_Closest_Elements.swift
class Solution {
    func findClosestElements(_ arr: [Int], _ k: Int, _ x: Int) -> [Int] {
        // sorted arr int
        var lst: [(Int, Int)] = []
        for num in arr {
            lst.append((abs(num - x), num))
        }

        lst.sort { ($0.0, $0.1) < ($1.0, $1.1) }

        let res = Array(lst.map { $0.1 }.prefix(k))

        return res.sorted()
    }
}

// Binary Search
class Solution {
    func findClosestElements(_ arr: [Int], _ k: Int, _ x: Int) -> [Int] {
        let n = arr.count
        guard n >= k else { return [] }

        var l = 0, r = n - k

        while l < r {
            let mid = (l + r) >> 1

            let leftDis = x - arr[mid]
            let rightDis = arr[mid + k] - x

            if leftDis > rightDis {
                l = mid + 1
            } else {
                r = mid
            }
        }

        return Array(arr[l ..< l + k])
    }
}

File: LeetCode/658.Find_K_Closest_Elements/658.Find_K_Closest_Elements.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


class Solution:
    def findClosestElements(self, arr: list[int], k: int, x: int) -> list[int]:
        n = len(arr)

        left, right = 0, n - k

        while left < right:
            mid = (left + right) // 2

            if x - arr[mid] > arr[mid + k] - x:
                left = mid + 1
            else:
                right = mid

        return arr[left : left + k]


class Solution2:
    def findClosestElements(self, arr: list[int], k: int, x: int) -> list[int]:

        # a[mid],a[mid+1],....,a[mid+k-1]
        #        a[mid+1],.....,a[mid+k-1], a[mid+k]
        # [a[mid]...  x. ... a[mid+k]]
        # x - a[mid] < a[mid+k] - x

        n = len(arr)

        left, right = 0, n - k - 1
        while left <= right:
            mid = (left + right) // 2

            if x - arr[mid] <= arr[mid + k] - x:
                right = mid - 1
            else:
                left = mid + 1

        return arr[left : left + k]
